File: src/main/java/no/nav/fo/forenkletdeploy/commits/StashCommitProvider.java
package no.nav.fo.forenkletdeploy.commits;

import lombok.Data;
import lombok.experimental.Accessors;
import no.nav.fo.forenkletdeploy.domain.ApplicationConfig;
import no.nav.fo.forenkletdeploy.domain.Commit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static no.nav.sbl.rest.RestUtils.withClient;

public class StashCommitProvider implements CommitProvider {
    private static final Logger LOGGER = LoggerFactory.getLogger(StashCommitProvider.class);
    private static final Integer LIMIT = 100;

    private Stream<StashCommit> getCommits(ApplicationConfig application, String fromTag, String toTag) {
        String url = getRestUriForRepo(application);
        LOGGER.info("Henter commits for " + application.name);
        try {
            return withClient(c -> c.target(url)
                    .queryParam("since", tagRef(fromTag))
                    .queryParam("until", tagRef(toTag))
                    .queryParam("limit", LIMIT)
                    .request()
                    .get(StashCommits.class)
            ).values.stream();
        } catch (Throwable t) {
            LOGGER.warn("Feil ved henting av commits for " + application.name);
            throw t;
        }
    }

    @Override
    public List<Commit> getCommitsForRelease(ApplicationConfig application, String fromTag, String toTag) {
        return this.getCommits(application, fromTag, toTag)
                .map(stashCommit -> mapToCommit(stashCommit, application))
                .collect(Collectors.toList());
    }

    public static String getRestUriForRepo(ApplicationConfig application) {
        Pattern pattern = Pattern.compile("7999/([a-zA-Z]+)/");
        Matcher matcher = pattern.matcher(application.getGitUrl());

        if(matcher.find()) {
            String project = matcher.group(1).toLowerCase();
            return String.format("http://stash.devillo.no/rest/api/1.0/projects/%s/repos/%s/commits", project, application.name);
        }
        return "";
    }

    private static String tagRef(String tag) {
        return "refs%2Ftags%2F" + tag;
    }

    private static Commit mapToCommit(StashCommit stashCommit, ApplicationConfig application) {
        return Commit.builder()
                .timestamp(stashCommit.committerTimestamp)
                .author(stashCommit.author.displayName)
                .message(stashCommit.message)
                .hash(stashCommit.id)
                .mergecommit(stashCommit.parents.size() > 1)
                .application(application.name)
                .build();
    }

    @Data
    @Accessors(fluent = true)
    private static class StashCommits {
        public boolean isLastPage;
        public int limit;
        public int nextPageStart;
        public int size;
        public int start;
        public List<StashCommit> values;
    }

    @Data
    @Accessors(fluent = true)
    private static class StashCommit {
        public Long authorTimestamp;
        public CommitPerson author;
        public Long committerTimestamp;
        public CommitPerson committer;
        public String displayId;
        public String id;
        public String message;
        public List<ParentCommit> parents;

    }

    @Data
    private static class CommitPerson {
        public boolean active;
        public String displayName;
        public String email;
        public int id;
        public String name;
        public String slug;
        public String type;
        public PersonLinks links;
    }

    @Data
    @Accessors(fluent = true)
    private static class PersonLinks {
        public List<PersonLink> self;
    }

    @Data
    @Accessors(fluent = true)
    private static class PersonLink {
        public String href;
    }

    @Data
    @Accessors(fluent = true)
    private static class ParentCommit {
        public String displayId;
        public String id;
    }
}

File: src/main/kotlin/no/nav/fo/forenkletdeploy/commits/StashCommitProvider.kt
package no.nav.fo.forenkletdeploy.commits

import com.github.javafaker.Faker
import no.nav.fo.forenkletdeploy.ApplicationConfig
import no.nav.fo.forenkletdeploy.Commit
import no.nav.fo.forenkletdeploy.util.Utils.withClient
import no.nav.fo.forenkletdeploy.util.stringToSeed
import org.slf4j.LoggerFactory
import java.util.*

interface IStashCommitProvider {
    fun getCommitsForRelease(application: ApplicationConfig, fromTag: String, toTag: String): List<Commit>
}

class StashCommitProvider: IStashCommitProvider {
    val LOG = LoggerFactory.getLogger(this.javaClass)
    val LIMIT = 1000

    override fun getCommitsForRelease(application: ApplicationConfig, fromTag: String, toTag: String): List<Commit> =
            getCommits(application, fromTag, toTag)
                    .map { Commit(
                            timestamp = it.committerTimestamp,
                            author = it.author.displayName ?: it.author.name,
                            url = getLinkUriForCommit(application, it.id),
                            message = it.message,
                            hash = it.id,
                            mergecommit = it.parents.size > 1,
                            application = application.name
                    ) }

    fun getCommits(application: ApplicationConfig, fromTag: String, toTag: String): List<StashCommit> =
            try {
                val url = getRestUriForRepo(application)
                LOG.info("Henter commits for ${application.name} ($fromTag -> $toTag) via $url")
                withClient(url)
                        .queryParam("since", tagRef(fromTag))
                        .queryParam("until", tagRef(toTag))
                        .queryParam("limit", LIMIT)
                        .request()
                        .get(StashCommits::class.java)
                        .values
            } catch (e: Throwable) {
                LOG.error("Feil ved henting av commits for ${application.name}", e)
                ArrayList()
            }

    private fun tagRef(tag: String) =
            if ("null".equals(tag, ignoreCase = true)) "" else "refs%2Ftags%2F$tag"
}

class MockStashCommitProvider: IStashCommitProvider {
    override fun getCommitsForRelease(application: ApplicationConfig, fromTag: String, toTag: String): List<Commit> {
        val faker = Faker(Random(stringToSeed(application.name + fromTag + toTag)))
        val numCommits = faker.number().numberBetween(1, 8)

        return (1..numCommits).map {
            Commit(
                    hash = faker.code().imei(),
                    application = application.name,
                    url = application.gitUrl,
                    message = getMessage(faker),
                    author = faker.zelda().character(),
                    timestamp = getMinutesAgo(it * 12),
                    mergecommit = false
            )
        }
    }

    fun getMessage(faker: Faker): String =
            faker.numerify("PUS-### ${faker.gameOfThrones().city()}\n\n${faker.chuckNorris().fact()}")

    fun getMinutesAgo(minutes: Int): Long =
            (System.currentTimeMillis()) - (minutes * 60 * 1000)
}

fun getRestUriForRepo(application: ApplicationConfig): String {
    val projectRegex = "7999/([a-zA-Z]+)/".toRegex()
    val project = projectRegex.find(application.gitUrl)?.groups?.get(1)?.value
    return "http://stash.devillo.no/rest/api/1.0/projects/$project/repos/${application.name}/commits"
}

fun getLinkUriForCommit(application: ApplicationConfig, commit: String): String {
    val projectRegex = "7999/([a-zA-Z]+)/".toRegex()
    val project = projectRegex.find(application.gitUrl)?.groups?.get(1)?.value
    return "http://stash.devillo.no/projects/$project/repos/${application.name}/commits/$commit"
}

data class StashCommits (
        val isLastPage: Boolean,
        val limit: Int,
        val nextPageStart: Int,
        val size: Int,
        val start: Int,
        val values: List<StashCommit>
)

data class StashCommit (
        val authorTimestamp: Long,
        val author: CommitPerson,
        val committerTimestamp: Long,
        val committer: CommitPerson,
        val displayId: String,
        val id: String,
        val message: String,
        val parents: List<ParentCommit>
)

data class CommitPerson (
        val active: Boolean,
        val displayName: String?,
        val email: String?,
        val id: Int,
        val name: String,
        val slug: String?,
        val type: String?,
        val links: PersonLinks?
)

data class PersonLinks (
        val self: List<PersonLink>
)

data class PersonLink (
        val href: String
)

data class ParentCommit (
        val displayId: String,
        val id: String
)